src/Database/Concerns/HasRelationships.php

```diff
@@ -279,520 +338,573 @@ protected function getRelationDefaults($type)
-                $relationObj = $this->$relationType($relation[0], $relation['public'], $relation['key'], $relationName);
+                $relation = $this->attachMany(
+                    $relatedClass,
+                    $definition['public'] ?? true,
+                    $definition['key'] ?? null,
+                    $definition['field'] ?? $relationName,
+                );
+                if (isset($definition['delete']) && $definition['delete'] === false) {
+                    $relation = $relation->notDependent();
+                }
                 break;
-
             case 'hasOneThrough':
+                $relation = $this->hasOneThrough(
+                    $relatedClass,
+                    $definition['through'],
+                    $definition['key'] ?? null,
+                    $definition['throughKey'] ?? null,
+                    $definition['otherKey'] ?? null,
+                    $definition['secondOtherKey'] ?? null,
+                );
+                break;
             case 'hasManyThrough':
-                $relation = $this->validateRelationArgs($relationName, ['key', 'throughKey', 'otherKey', 'secondOtherKey'], ['through']);
-                $relationObj = $this->$relationType($relation[0], $relation['through'], $relation['key'], $relation['throughKey'], $relation['otherKey'], $relation['secondOtherKey'], $relationName);
+                $relation = $this->hasManyThrough(
+                    $relatedClass,
+                    $definition['through'],
+                    $definition['key'] ?? null,
+                    $definition['throughKey'] ?? null,
+                    $definition['otherKey'] ?? null,
+                    $definition['secondOtherKey'] ?? null,
+                );
                 break;
-
             default:
-                throw new InvalidArgumentException(sprintf("There is no such relation type known as '%s' on model '%s'.", $relationType, get_called_class()));
+                throw new InvalidArgumentException(
+                    sprintf(
+                        'There is no such relation type known as \'%s\' on model \'%s\'.',
+                        $relationType,
+                        get_called_class()
+                    )
+                );
         }
 
-        return $relationObj;
-    }
-
-    /**
-     * Validate relation supplied arguments.
-     */
-    protected function validateRelationArgs($relationName, $optional, $required = [])
-    {
-        $relation = $this->getRelationDefinition($relationName);
+        // Add relation name
+        $relation->setRelationName($relationName);
+
+        // Add dependency, if required
+        if (
+            in_array(
+                \Winter\Storm\Database\Relations\Concerns\CanBeDependent::class,
+                class_uses_recursive($relation)
+            )
+            && (($definition['delete'] ?? false) === true)
```

Comment:
Should we check the definition configs before making calls to reflection?

Reply:
Fixed.